File: frontend/be_elite/lib/models/Session/SessionCardDto/session_card_dto.dart
import 'block_dto.dart';

class SessionCardDto {
	String? date;
	int? sessionNumber;
	String? programId;
	String? weekName;
	int? weekNumber;
	List<Block>? blocks;
	String? title;
	String? subtitle;
	int? sameDaySessionNumber;

	SessionCardDto({
		this.date, 
		this.sessionNumber, 
		this.programId, 
		this.weekName, 
		this.weekNumber, 
		this.blocks, 
		this.title, 
		this.subtitle, 
		this.sameDaySessionNumber, 
	});

	factory SessionCardDto.fromJson(Map<String, dynamic> json) => SessionCardDto(
				date: json['date'] as String?,
				sessionNumber: json['sessionNumber'] as int?,
				programId: json['programId'] as String?,
				weekName: json['weekName'] as String?,
				weekNumber: json['weekNumber'] as int?,
				blocks: (json['blocks'] as List<dynamic>?)
						?.map((e) => Block.fromJson(e as Map<String, dynamic>))
						.toList(),
				title: json['title'] as String?,
				subtitle: json['subtitle'] as String?,
				sameDaySessionNumber: json['sameDaySessionNumber'] as int?,
			);

	Map<String, dynamic> toJson() => {
				'date': date,
				'sessionNumber': sessionNumber,
				'programId': programId,
				'weekName': weekName,
				'weekNumber': weekNumber,
				'blocks': blocks?.map((e) => e.toJson()).toList(),
				'title': title,
				'subtitle': subtitle,
				'sameDaySessionNumber': sameDaySessionNumber,
			};
}

File: backend/BeElite/src/main/java/com/salesianos/triana/BeElite/dto/Session/SessionCardDto.java
package com.salesianos.triana.BeElite.dto.Session;

import com.salesianos.triana.BeElite.dto.Block.BlockDto;
import com.salesianos.triana.BeElite.model.AthleteSession;
import com.salesianos.triana.BeElite.model.Session;
import lombok.Builder;

import java.time.LocalDate;
import java.util.List;
import java.util.UUID;

@Builder
public record SessionCardDto(LocalDate date,
                                Long sessionNumber,
                                UUID programId,
                                String weekName,
                                Long weekNumber,
                                List<BlockDto> blocks,
                                String title,
                                String subtitle,
                                int sameDaySessionNumber) {

    public static SessionCardDto of(AthleteSession athleteSession){
        Session s = athleteSession.getSession();

        return SessionCardDto.builder()
                .date(s.getDate())
                .sessionNumber(s.getId().getSession_number())
                .programId(s.getId().getWeek_id().getProgram_id())
                .weekName(s.getId().getWeek_id().getWeek_name())
                .weekNumber(s.getId().getWeek_id().getWeek_number())
                .blocks(athleteSession.getAthleteBlocks() != null ?
                        athleteSession.getAthleteBlocks().stream().map(BlockDto::of).toList() :
                        List.of())
                .title(s.getTitle())
                .subtitle(s.getSubtitle())
                .sameDaySessionNumber(s.getSame_day_session_number())
                .build();
    }
}
